source/src/libs/xpdf/doc/pdfinfo.1
.\" Copyright 1999-2003 Glyph & Cog, LLC
.TH pdfinfo 1 "24 March 2003"
.SH NAME
pdfinfo \- Portable Document Format (PDF) document information
extractor (version 2.02)
.SH SYNOPSIS
.B pdfinfo
[options]
.RI [ PDF-file ]
.SH DESCRIPTION
.B Pdfinfo
prints the contents of the \'Info' dictionary (plus some other useful
information) from a Portable Document Format (PDF) file.
.PP
The \'Info' dictionary contains the following values:
.PP
.RS
title
.RE
.RS
subject
.RE
.RS
keywords
.RE
.RS
author
.RE
.RS
creator
.RE
.RS
producer
.RE
.RS
creation date
.RE
.RS
modification date
.RE
.PP
In addition, the following information is printed:
.PP
.RS
page count
.RE
.RS
encrypted flag (yes/no)
.RE
.RS
print and copy permissions (if encrypted)
.RE
.RS
linearized (yes/no)
.RE
.SH CONFIGURATION FILE
Pdfinfo reads a configuration file at startup.  It first tries to find
the user's private config file, ~/.xpdfrc.  If that doesn't exist, it
looks for a system-wide config file, typically /usr/local/etc/xpdfrc
(but this location can be changed when pdfinfo is built).  See the
.BR xpdfrc (5)
man page for details.
.SH OPTIONS
Many of the following options can be set with configuration file
commands.  These are listed in square brackets with the description of
the corresponding command line option.
.TP
.BI \-meta
Prints document-level metadata.  (This is the "Metadata" stream from
the PDF file's Catalog object.)
.TP
.BI \-enc " encoding-name"
Sets the encoding to use for text output.  The
.I encoding\-name
must be defined with the unicodeMap command (see
.BR xpdfrc (5)).
This defaults to "Latin1" (which is a built-in encoding).
.RB "[config file: " textEncoding ]
.TP
.BI \-opw " password"
Specify the owner password for the PDF file.  Providing this will
bypass all security restrictions.
.TP
.BI \-upw " password"
Specify the user password for the PDF file.
.TP
.BI \-cfg " config-file"
Read
.I config-file
in place of ~/.xpdfrc or the system-wide config file.
.TP
.B \-v
Print copyright and version information.
.TP
.B \-h
Print usage information.
.RB ( \-help
and
.B \-\-help
are equivalent.)
.SH EXIT CODES
The Xpdf tools use the following exit codes:
.TP
0
No error.
.TP
1
Error opening a PDF file.
.TP
2
Error opening an output file.
.TP
3
Error related to PDF permissions.
.TP
99
Other error.
.SH AUTHOR
The pdfinfo software and documentation are copyright 1996-2003 Glyph &
Cog, LLC.
.SH "SEE ALSO"
.BR xpdf (1),
.BR pdftops (1),
.BR pdftotext (1),
.BR pdffonts (1),
.BR pdftopbm (1),
.BR pdfimages (1),
.BR xpdfrc (5)
.br
.B http://www.foolabs.com/xpdf/


www/NEWS
-------------------------------------------------
pdfTeX 3.141592-1.40.7 was released on 2008-02-17
-------------------------------------------------
- Bugfix:
  - PDFs > 2GiB (LFS) were still broken on 32-bit systems

-------------------------------------------------
pdfTeX 3.141592-1.40.6 was released on 2008-02-04
-------------------------------------------------
- Bugfixes:
  - inclusion of encrypted PDFs was broken (861)
  - PDFs > 2GiB (LFS) would be broken, even on 64-bit systems. Now the basic
    support for LFS is there, you only need to supply the right compiler flags
    (this is not yet done automatically). See build.sh for hints on compiling.
  - Page groups in pdfs (see the discussion in the notes vor v1.10a) are now
    handled correctly (741)
  - When including transparent PNGs pdftex now automatically adds a Page Group
    (896)
  - broken map files would lead to core dumps
  - unusual TrueType fonts would lead to an assertion error (874)
  - the null object in the object table was never initialized
  - included PDFs with /PieceInfo lost their /LastModified date
  - an empty colorstack crashed pdftex (897)
- pdfTeX uses libpng 1.2.24
- pdfTeX uses xpdf 3.02pl2

-------------------------------------------------
pdfTeX 3.141592-1.40.5 was released on 2007-07-31
-------------------------------------------------
- Bugfixes: 
  - calling 'pdftex %' would crash (830)
  - fflush the recorder file
- pdfTeX uses xpdf 3.02pl1

-------------------------------------------------
pdfTeX 3.141592-1.40.4 was released on 2007-07-11
-------------------------------------------------
- Bugfixes: 
  - It was impossible to set the font Flags to 4 in map entries (762)
  - The font Flags of non-embedded fonts were wrong (794)
    Now pdfTeX uses the Flags from map entries (if found) or guesses values for
    the base14 fonts. It also warns when you not-embedd a font without Flags
  - Subsetted fonts still had their private UniqueID copied (774)
  - \pdfunescapehex returned a wrong result if the last character was not a
    hexadecimal digit and the last hex pair was incomplete (777)
  - Copying a box with \pdfstartlink inside produced the same /Link annotation
    object multiply, with same object number and identical coordinates on the
    page (799)
  - When trying to use a font without TFM, without map entry, and no PK font
    available, a broken PDF was produced. Now no PDF is generated and the error
    messages have been improved (721)
  - Problems with vf fonts (722)
  - With pdf inclusion the font descriptor could get mixed up and miss some
    required keys (810)
  - Font expansion led to wrong log output of \kerns (736)
  - \pdfprimitive was broken (see
    http://www.ntg.nl/pipermail/ntg-pdftex/2007-April/002860.html)
  - The -recorder option of web2c didn't record fonts, included images and some
    other files (772)
  - Using \pdfximage in IniTeX before \dump caused a crash when the format was
    used (824)
  - \letterspacefont had some problems (712)
  - \letterspacefont and \pdffontexpand had some problems (747)
- Improvements: 
  - pdf compression is now a bit faster (patch 726)
  - the behaviour of TeX when it asks for a file name has been improved (756)
- pdfTeX uses libpng 1.2.18
- pdfTeX uses xpdf 3.02

-------------------------------------------------
pdfTeX 3.141592-1.40.3 was released on 2007-02-04
-------------------------------------------------
- bugfix: document which used type 1 and truetype or opentype fonts resulted in
  broken pdfs (729)

-------------------------------------------------
pdfTeX 3.141592-1.40.2 was released on 2007-01-31
-------------------------------------------------
- bugfix: maplines starting with = would not work as advertised
- bugfix: xpdf would complain that PDF 1.7 is too new

-------------------------------------------------
pdfTeX 3.141592-1.40.1 was released on 2007-01-08
-------------------------------------------------
- bugfix: the shell escape function was a bit broken
- bugfix: \pdffontattr gave confusing warnings
- allow JBIG2 images only with PDF >= 1.4
- allow \pdfglyphtounicode in dvi mode for format generation
- pdfTeX uses libpng 1.2.15

-------------------------------------------------
pdfTeX 3.141592-1.40.0 was released on 2007-01-01
-------------------------------------------------
- pdfetex (as a separate binary/program) is gone; all extensions are now in
  pdftex. The make files still know about the target pdfetex, but it's just a
  copy of pdftex
- The sources of TeX, eTeX and pdfTeX have been merged into one file pdftex.web
  and one change file pdftex.ch. The sources of TeX and eTeX distributed are
  just there for reference
- New features:
  - pdfTeX can now generate PDF 1.5 object streams, which leads to smaller
    PDFs. This is controlled by setting \pdfobjcompresslevel to a value >0 (3
    compress everything, 2 don't compress /Info, 1 also don't compress embedded
    PDFs) and works only with PDF >=1.5
  - pdfTeX now supports another image file format: JBIG2 files (.jbig2)
  - the memory allocated for objects and destination names now grows
    dynamically
  - pdfTeX now generates smaller PDFs since for char positioning the width of
    chars (/Widths) is stored with more precision and so the chars must rarely
    be positioned separately
  - the PK resolution is now taken from the "pk_dpi" parameter in texmf.cnf if
    it has not been set in the format file or by the user
  - pdfTeX now removes the generated pdf in case of a fatal error
  - pdfTeX now sets /ModDate and /Trapped in the InfoDict by default (the
    values can be overridden with \pdfinfo)
  - the format of warnings and errors issued by pdfTeX has been changed and
    unified
  - the output of -version has been extended and now includes information about
    the libraries actually used, which can be handy when using shared libraries
  - rules and their positioning on the page are now inside a qQ group
- New primitives:

source/src/libs/xpdf/xpdf/pdfinfo.cc
//========================================================================
//
// pdfinfo.cc
//
// Copyright 1998-2003 Glyph & Cog, LLC
//
//========================================================================

#include <aconf.h>
#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include <string.h>
#include <time.h>
#include <math.h>
#include "parseargs.h"
#include "GString.h"
#include "gmem.h"
#include "GlobalParams.h"
#include "Object.h"
#include "Stream.h"
#include "Array.h"
#include "Dict.h"
#include "XRef.h"
#include "Catalog.h"
#include "Page.h"
#include "PDFDoc.h"
#include "CharTypes.h"
#include "UnicodeMap.h"
#include "Error.h"
#include "config.h"

static void printInfoString(Dict *infoDict, char *key, char *text,
			    UnicodeMap *uMap);
static void printInfoDate(Dict *infoDict, char *key, char *text);

static GBool printMetadata = gFalse;
static char textEncName[128] = "";
static char ownerPassword[33] = "";
static char userPassword[33] = "";
static char cfgFileName[256] = "";
static GBool printVersion = gFalse;
static GBool printHelp = gFalse;

static ArgDesc argDesc[] = {
  {"-meta",   argFlag,     &printMetadata,    0,
   "print the document metadata (XML)"},
  {"-enc",    argString,   textEncName,    sizeof(textEncName),
   "output text encoding name"},
  {"-opw",    argString,   ownerPassword,  sizeof(ownerPassword),
   "owner password (for encrypted files)"},
  {"-upw",    argString,   userPassword,   sizeof(userPassword),
   "user password (for encrypted files)"},
  {"-cfg",        argString,      cfgFileName,    sizeof(cfgFileName),
   "configuration file to use in place of .xpdfrc"},
  {"-v",      argFlag,     &printVersion,  0,
   "print copyright and version info"},
  {"-h",      argFlag,     &printHelp,     0,
   "print usage information"},
  {"-help",   argFlag,     &printHelp,     0,
   "print usage information"},
  {"--help",  argFlag,     &printHelp,     0,
   "print usage information"},
  {"-?",      argFlag,     &printHelp,     0,
   "print usage information"},
  {NULL}
};

int main(int argc, char *argv[]) {
  PDFDoc *doc;
  GString *fileName;
  GString *ownerPW, *userPW;
  UnicodeMap *uMap;
  Object info;
  double w, h, wISO, hISO;
  FILE *f;
  GString *metadata;
  GBool ok;
  int exitCode;
  int i;

  exitCode = 99;

  // parse args
  ok = parseArgs(argDesc, &argc, argv);
  if (!ok || argc != 2 || printVersion || printHelp) {
    fprintf(stderr, "pdfinfo version %s\n", xpdfVersion);
    fprintf(stderr, "%s\n", xpdfCopyright);
    if (!printVersion) {
      printUsage("pdfinfo", "<PDF-file>", argDesc);
    }
    goto err0;
  }
  fileName = new GString(argv[1]);

  // read config file
  globalParams = new GlobalParams(cfgFileName);
  if (textEncName[0]) {
    globalParams->setTextEncoding(textEncName);
  }

  // get mapping to output encoding
  if (!(uMap = globalParams->getTextEncoding())) {
    error(-1, "Couldn't get text encoding");
    delete fileName;
    goto err1;
  }

  // open PDF file
  if (ownerPassword[0]) {
    ownerPW = new GString(ownerPassword);
  } else {
    ownerPW = NULL;
  }
  if (userPassword[0]) {
    userPW = new GString(userPassword);
  } else {
    userPW = NULL;
  }
  doc = new PDFDoc(fileName, ownerPW, userPW);
  if (userPW) {
    delete userPW;
  }
  if (ownerPW) {
    delete ownerPW;
  }
  if (!doc->isOk()) {
    exitCode = 1;
    goto err2;
  }

  // print doc info
  doc->getDocInfo(&info);
  if (info.isDict()) {
    printInfoString(info.getDict(), "Title",        "Title:        ", uMap);
    printInfoString(info.getDict(), "Subject",      "Subject:      ", uMap);
    printInfoString(info.getDict(), "Keywords",     "Keywords:     ", uMap);
    printInfoString(info.getDict(), "Author",       "Author:       ", uMap);
    printInfoString(info.getDict(), "Creator",      "Creator:      ", uMap);
    printInfoString(info.getDict(), "Producer",     "Producer:     ", uMap);
    printInfoDate(info.getDict(),   "CreationDate", "CreationDate: ");
    printInfoDate(info.getDict(),   "ModDate",      "ModDate:      ");
  }
  info.free();

  // print tagging info
  printf("Tagged:       %s\n",
	 doc->getStructTreeRoot()->isDict() ? "yes" : "no");

  // print page count
  printf("Pages:        %d\n", doc->getNumPages());

  // print encryption info
  printf("Encrypted:    ");
  if (doc->isEncrypted()) {
    printf("yes (print:%s copy:%s change:%s addNotes:%s)\n",
	   doc->okToPrint(gTrue) ? "yes" : "no",
	   doc->okToCopy(gTrue) ? "yes" : "no",
	   doc->okToChange(gTrue) ? "yes" : "no",
	   doc->okToAddNotes(gTrue) ? "yes" : "no");
  } else {
    printf("no\n");
  }

  // print page size
  if (doc->getNumPages() >= 1) {
    w = doc->getPageWidth(1);
    h = doc->getPageHeight(1);
    printf("Page size:    %g x %g pts", w, h);
    if ((fabs(w - 612) < 0.1 && fabs(h - 792) < 0.1) ||
	(fabs(w - 792) < 0.1 && fabs(h - 612) < 0.1)) {
      printf(" (letter)");
    } else {
      hISO = sqrt(sqrt(2.0)) * 7200 / 2.54;
      wISO = hISO / sqrt(2.0);
      for (i = 0; i <= 6; ++i) {
	if ((fabs(w - wISO) < 1 && fabs(h - hISO) < 1) ||
	    (fabs(w - hISO) < 1 && fabs(h - wISO) < 1)) {
	  printf(" (A%d)", i);
	  break;
	}
	hISO = wISO;
	wISO /= sqrt(2.0);
      }
    }
    printf("\n");
  } 

  // print file size
#ifdef VMS
  f = fopen(fileName->getCString(), "rb", "ctx=stm");
#else
  f = fopen(fileName->getCString(), "rb");
#endif
  if (f) {
#if HAVE_FSEEKO
    fseeko(f, 0, SEEK_END);
    printf("File size:    %u bytes\n", (Guint)ftello(f));
#elif HAVE_FSEEK64
    fseek64(f, 0, SEEK_END);
    printf("File size:    %u bytes\n", (Guint)ftell64(f));
#else
    fseek(f, 0, SEEK_END);
    printf("File size:    %d bytes\n", (int)ftell(f));
#endif
    fclose(f);
  }

  // print linearization info
  printf("Optimized:    %s\n", doc->isLinearized() ? "yes" : "no");

  // print PDF version
  printf("PDF version:  %.1f\n", doc->getPDFVersion());

  // print the metadata
  if (printMetadata && (metadata = doc->readMetadata())) {
    fputs("Metadata:\n", stdout);
    fputs(metadata->getCString(), stdout);
    fputc('\n', stdout);
    delete metadata;
  }

  exitCode = 0;

  // clean up
 err2:
  uMap->decRefCnt();
  delete doc;
 err1:
  delete globalParams;
 err0:

  // check for memory leaks
  Object::memCheck(stderr);
  gMemReport(stderr);

  return exitCode;
}

static void printInfoString(Dict *infoDict, char *key, char *text,
			    UnicodeMap *uMap) {
  Object obj;
  GString *s1;
  GBool isUnicode;
  Unicode u;
  char buf[8];
  int i, n;

  if (infoDict->lookup(key, &obj)->isString()) {
    fputs(text, stdout);
    s1 = obj.getString();
    if ((s1->getChar(0) & 0xff) == 0xfe &&
	(s1->getChar(1) & 0xff) == 0xff) {
      isUnicode = gTrue;
      i = 2;
    } else {
      isUnicode = gFalse;
      i = 0;
    }
    while (i < obj.getString()->getLength()) {
      if (isUnicode) {
	u = ((s1->getChar(i) & 0xff) << 8) |
	    (s1->getChar(i+1) & 0xff);
	i += 2;
      } else {
	u = s1->getChar(i) & 0xff;
	++i;
      }
      n = uMap->mapUnicode(u, buf, sizeof(buf));
      fwrite(buf, 1, n, stdout);
    }
    fputc('\n', stdout);
  }
  obj.free();
}

static void printInfoDate(Dict *infoDict, char *key, char *text) {
  Object obj;
  char *s;
  int year, mon, day, hour, min, sec;
  struct tm tmStruct;
  char buf[256];

  if (infoDict->lookup(key, &obj)->isString()) {
    fputs(text, stdout);
    s = obj.getString()->getCString();
    if (s[0] == 'D' && s[1] == ':') {
      s += 2;
    }
    if (sscanf(s, "%4d%2d%2d%2d%2d%2d",
	       &year, &mon, &day, &hour, &min, &sec) == 6) {
      tmStruct.tm_year = year - 1900;
      tmStruct.tm_mon = mon - 1;
      tmStruct.tm_mday = day;
      tmStruct.tm_hour = hour;
      tmStruct.tm_min = min;
      tmStruct.tm_sec = sec;
      tmStruct.tm_wday = -1;
      tmStruct.tm_yday = -1;
      tmStruct.tm_isdst = -1;
      // compute the tm_wday and tm_yday fields
      if (mktime(&tmStruct) != (time_t)-1 &&
	  strftime(buf, sizeof(buf), "%c", &tmStruct)) {
	fputs(buf, stdout);
      } else {
	fputs(s, stdout);
      }
    } else {
      fputs(s, stdout);
    }
    fputc('\n', stdout);
  }
  obj.free();
}
